File: code/pipeline-rust/src/thread/telemetry.rs
use std::{
    fs::File,
    io::{Read, Write},
    sync::{
        atomic::Ordering,
        mpsc::{Receiver, SyncSender},
    },
    thread::{self, JoinHandle},
    time::{Duration, Instant},
};

use anyhow::{Context, Result};
use hdrhistogram::Histogram;

use crate::os::{ALLOCATED_BYTES, ALLOCATION_COUNT, FREED_BYTES};

/// A telemetry epoch that contains six `HdrHistogram` instances for recording
/// latency measurements in nanoseconds. Each histogram tracks latency for a
/// specific stage of the inference pipeline, as well as the total latency.
pub struct TelemetryEpoch {
    /// The histograms for recording latency measurements in nanoseconds for
    /// each stage of the inference pipeline, as well as the total latency.
    latency_nanos: [Histogram<u64>; 6],

    /// The total number of frames lapped during the epoch.
    lapped_frames: u64,

    /// The total number of frames dropped during the epoch.
    dropped_frames: u64,

    /// The total number of bytes allocated during the epoch.
    allocated_bytes: usize,

    /// The total number of allocations made during the epoch.
    allocation_count: usize,

    /// The total number of bytes freed during the epoch.
    freed_bytes: usize,

    /// The total number of allocated RSS bytes during the epoch.
    pub rss_bytes: usize,

    /// The total number of allocated fordblks bytes during the epoch.
    pub fordblks_bytes: usize,

    /// A flag indicating whether the telemetry thread should terminate.
    terminated: bool,
}

impl TelemetryEpoch {
    /// The index of the histogram for unbounded queue wait latency.
    pub const UNBOUNDED_QUEUE_WAIT: usize = 0;

    /// The index of the histogram for idiomatic queue wait latency.
    #[allow(unused)]
    pub const IDIOMATIC_QUEUE_WAIT: usize = 1;

    /// The index of the histogram for inference latency.
    #[allow(unused)]
    pub const INFERENCE_EXEC: usize = 2;

    /// The index of the histogram for multi-producer single-consumer (MPSC)
    /// queue wait latency.
    #[allow(unused)]
    pub const MPSC_WAIT: usize = 3;

    /// The index of the histogram for fusion latency.
    #[allow(unused)]
    pub const FUSION_EXEC: usize = 4;

    /// The index of the histogram for total latency.
    pub const TOTAL_LATENCY: usize = 5;

    /// The number of latency measures tracked in an epoch.
    pub const NUM_LATENCY_MEASURES: usize = 6;

    /// Creates a new `TelemetryEpoch` with six HdrHistogram instances for
    /// recording latency measurements in nanoseconds. Each histogram is
    /// initialised to track values from 1 nanosecond to 60 seconds with 3
    /// significant figures of precision.
    ///
    /// Returns a `Result` containing the newly created `TelemetryEpoch`, or an
    /// error if any of the histograms could not be initialised.
    pub fn try_new() -> Result<Self> {
        let unbounded_queue_wait = Self::create_histogram().context(
            "Failed to create HdrHistogram for unbounded queue wait latency",
        )?;

        let idiomatic_queue_wait = Self::create_histogram().context(
            "Failed to create HdrHistogram for idiomatic queue wait latency",
        )?;

        let inference_exec = Self::create_histogram().context(
            "Failed to create HdrHistogram for inference exec latency",
        )?;

        let mpsc_wait = Self::create_histogram()
            .context("Failed to create HdrHistogram for MPSC wait latency")?;

        let fusion_exec = Self::create_histogram()
            .context("Failed to create HdrHistogram for fusion exec latency")?;

        let total_latency = Self::create_histogram()
            .context("Failed to create HdrHistogram for total latency")?;

        Ok(Self {
            latency_nanos: [
                unbounded_queue_wait,
                idiomatic_queue_wait,
                inference_exec,
                mpsc_wait,
                fusion_exec,
                total_latency,
            ],
            lapped_frames: 0,
            dropped_frames: 0,
            allocated_bytes: 0,
            allocation_count: 0,
            freed_bytes: 0,
            rss_bytes: 0,
            fordblks_bytes: 0,
            terminated: false,
        })
    }

    /// Resets the histograms and allocation statistics for the epoch.
    fn reset(&mut self) {
        for histogram in self.latency_nanos.iter_mut() {
            histogram.reset();
        }

        self.lapped_frames = 0;
        self.dropped_frames = 0;
        self.allocated_bytes = 0;
        self.allocation_count = 0;
        self.freed_bytes = 0;
        self.rss_bytes = 0;
        self.fordblks_bytes = 0;
    }

    /// Creates a new HdrHistogram instance for recording latency measurements
    /// in nanoseconds. The histogram is initialised to track values from 1
    /// nanosecond to 60 seconds with 3 significant figures of precision.
    ///
    /// Returns a `Result` containing the newly created `Histogram<u64>`, or an
    /// error if the histogram could not be initialised.
    fn create_histogram() -> Result<Histogram<u64>> {
        Histogram::<u64>::new_with_bounds(1, 60_000_000_000, 3)
            .context("Failed to create HdrHistogram")
    }
}

/// A CSV file writer for telemetry data.
struct Csv {
    file: File,
}

impl Csv {
    /// Creates a new `Csv` writer for the specified filename. The writer is
    /// initialised with a header row containing the names of the latency
    /// percentiles for each stage of the inference pipeline, as well as the
    /// total latency.
    ///
    /// * `filename` - The name of the CSV file to write to.
    ///
    /// Returns a `Result` containing the newly created `Csv`, or an error if
    /// the file could not be created or the header row could not be written.
    fn try_new(filename: impl Into<String>) -> Result<Self> {
        let filename = filename.into();
        let mut file = File::create(&filename).with_context(|| {
            format!("Failed to create CSV telemtry file '{}'", filename)
        })?;

        let header = b"unbounded_wait_p50,unbounded_wait_p99,unbounded_wait_p99_9,unbounded_wait_max,\
            idiomatic_wait_p50,idiomatic_wait_p99,idiomatic_wait_p99_9,idiomatic_wait_max,\
            inference_exec_p50,inference_exec_p99,inference_exec_p99_9,inference_exec_max,\
            mpsc_wait_p50,mpsc_wait_p99,mpsc_wait_p99_9,mpsc_wait_max,\
            fusion_exec_p50,fusion_exec_p99,fusion_exec_p99_9,fusion_exec_max,\
            total_latency_p50,total_latency_p99,total_latency_p99_9,total_latency_max,\
            lapped_frames,dropped_frames,allocated_bytes,allocation_count,\
            freed_bytes,rss_bytes,fordblks_bytes\n";

        file.write_all(header).expect("Failed to write CSV header");

        Ok(Self { file })
    }

    /// Writes a telemetry record to the CSV file.
    ///
    /// * `epoch` - The telemetry epoch containing the latency measurements to
    /// be written.
    ///
    /// Returns a `Result` indicating whether the record was successfully
    /// written, or an error if the operation failed.
    fn write_record(&mut self, epoch: &TelemetryEpoch) -> Result<()> {
        const BUFFER_SIZE: usize = 1024;

        let mut buf = [0u8; BUFFER_SIZE];
        let mut cursor = &mut buf[..];
        let mut itoa_buf = itoa::Buffer::new();

        let mut append = |val: u64| {
            let s = itoa_buf.format(val);
            cursor.write_all(s.as_bytes()).unwrap();
            cursor.write_all(b",").unwrap();
        };

        for histogram in epoch.latency_nanos.iter() {
            append(histogram.value_at_quantile(0.5));
            append(histogram.value_at_quantile(0.99));
            append(histogram.value_at_quantile(0.999));
            append(histogram.max());
        }

        append(epoch.lapped_frames);
        append(epoch.dropped_frames);
        append(epoch.allocated_bytes as u64);
        append(epoch.allocation_count as u64);
        append(epoch.freed_bytes as u64);
        append(epoch.rss_bytes as u64);
        append(epoch.fordblks_bytes as u64);

        // Replace final comma with newline.
        let written = BUFFER_SIZE - cursor.len();
        buf[written - 1] = b'\n';

        self.file
            .write_all(&buf[..written])
            .context("Failed to write telemetry record to CSV")?;

        Ok(())
    }
}

/// Records latency measurements into an epoch buffer. Used by the inference
/// thread to publish the results to the telemetry thread for processing, and to
/// receive a fresh buffer for the next epoch.
pub struct TelemetryWriter {
    /// The sender channel for publishing the current epoch to the telemetry
    /// thread for processing.
    sender: SyncSender<TelemetryEpoch>,

    /// The receiver channel for receiving a fresh epoch from the telemetry
    /// thread.
    receiver: Receiver<TelemetryEpoch>,

    /// The timestamp of the last buffer swap, used to determine when to next
    /// swap the buffers.
    last_swap: Instant,

    /// The number of frames lapped since the call to record, used to calculate
    /// the number of lapped frames for the current epoch.
    last_lapped_frames: u64,

    /// The number of frames dropped since the call to record, used to
    /// calculate the number of dropped frames for the current epoch.
    last_dropped_frames: u64,

    /// The currently active telemetry epoch for recording latency measurements.
    current_epoch: TelemetryEpoch,
}

impl TelemetryWriter {
    /// The interval at which the active epoch buffer is swapped with a fresh
    /// buffer received from the telemetry thread.
    const SWAP_INTERVAL: Duration = Duration::from_secs(1);

    /// Constructs a new `TelemetryWriter` with the specified sender and
    /// receiver channels for double-buffered histogram processing.
    ///
    /// * `sender` - The sender channel for publishing the populated epoch to
    ///   the telemetry thread for processing.
    /// * `receiver` - The receiver channel for receiving a fresh epoch buffer
    ///   from the telemetry thread.
    ///
    /// Returns a `Result` containing the newly created `TelemetryWriter`, or an
    /// error if the histogram buffer could not be initialised
    pub fn try_new(
        sender: SyncSender<TelemetryEpoch>,
        receiver: Receiver<TelemetryEpoch>,
    ) -> Result<Self> {
        let current_epoch = receiver
            .recv()
            .context("Failed to receive initial telemetry epoch")?;

        Ok(Self {
            current_epoch,
            sender,
            receiver,
            last_swap: Instant::now(),
            last_lapped_frames: 0,
            last_dropped_frames: 0,
        })
    }

    /// Records latency measurements into the currently active telemetry epoch.
    /// If at least one second has elapsed since the last buffer swap, the
    /// active buffer is swapped with a fresh buffer received from the telemetry
    /// thread.
    ///
    /// * `ts` - An array of six timestamps in nanoseconds, representing the
    ///   start and end times of each stage of the inference pipeline, as well
    ///   as the total latency.
    /// * `lapped_frames` - The total number of frames lapped.
    /// * `dropped_frames` - The total number of frames dropped.
    ///
    /// Returns a `Result` indicating whether the latency measurements were
    /// successfully recorded, or an error if the operation failed.
    pub fn record(
        &mut self,
        ts: [u64; 6],
        lapped_frames: u64,
        dropped_frames: u64,
    ) -> Result<()> {
        if self.last_swap.elapsed() >= Self::SWAP_INTERVAL {
            self.swap_buffers()?;
        }

        for idx in 0..(TelemetryEpoch::NUM_LATENCY_MEASURES - 1) {
            let nanos = ts[idx + 1].saturating_sub(ts[idx]).max(1);
            self.current_epoch.latency_nanos[idx]
                .record(nanos)
                .context("Failed to record latency")?;
        }

        let total_nanos = ts[TelemetryEpoch::TOTAL_LATENCY]
            .saturating_sub(ts[TelemetryEpoch::UNBOUNDED_QUEUE_WAIT]);
        self.current_epoch.latency_nanos[TelemetryEpoch::TOTAL_LATENCY]
            .record(total_nanos)
            .context("Failed to record total latency")?;

        let newly_lapped =
            lapped_frames.saturating_sub(self.last_lapped_frames);
        self.current_epoch.lapped_frames += newly_lapped;
        self.last_lapped_frames = lapped_frames;

        let newly_dropped =
            dropped_frames.saturating_sub(self.last_dropped_frames);
        self.current_epoch.dropped_frames += newly_dropped;
        self.last_dropped_frames = dropped_frames;

        Ok(())
    }

    /// Swaps the active histogram buffer with the cleared buffer received from
    /// the background thread. The populated buffer is sent to the background
    /// thread for processing, and the timestamp of the last swap is updated.
    fn swap_buffers(&mut self) -> Result<()> {
        // Try to receive the fresh epoch from the telemetry thread. If the
        // channel is empty, we skip the swap and continue recording into the
        // current epoch.
        if let Ok(mut epoch) = self.receiver.try_recv() {
            self.last_swap = Instant::now();

            // Using `std::mem::swap` to swap the populated epoch with the fresh
            // epoch, avoiding the need for cloning or moving the data.
            std::mem::swap(&mut self.current_epoch, &mut epoch);

            // Send the completed epoch to the telemetry thread for processing.
            self.sender
                .try_send(epoch)
                .expect("Failed to send completed epoch to telemetry thread")
        }

        Ok(())
    }

    /// Signals the telemetry thread to terminate by setting the `terminated`
    /// flag in the current epoch and sending it to the telemetry thread.
    pub fn terminate(&mut self) -> Result<()> {
        self.current_epoch.terminated = true;
        self.swap_buffers()
    }
}

/// Spawns a background thread for processing telemetry data from a
/// `TelemetryWriter`. The telemtry thread receives populated epoch buffers from
/// the `TelemetryWriter`, calculates latency percentiles, and saves the results
/// to a CSV file for later analysis.
///
/// * `stream_name` - The name of the stream to be used for telemetry and thread
///   identification.
/// * `sender` - The sender channel for publishing the current epoch to the
///   telemetry thread for processing.
/// * `receiver` - The receiver channel for receiving a fresh epoch buffer from
///   the telemetry thread.
///
/// Returns a `Result` containing the join handle for the spawned thread, or an
/// error if the thread could not be spawned.
pub fn spawn_telemetry_thread(
    stream_name: impl Into<String>,
    sender: SyncSender<TelemetryEpoch>,
    receiver: Receiver<TelemetryEpoch>,
) -> Result<JoinHandle<()>> {
    let stream_name = stream_name.into();

    // The `TelemetryEpoch` is used to record measurements, and is what is
    // communicated between the inference thread and the telemetry thread. Three
    // epochs are created:
    // * One will be used by the inference thread to record measurements.
    // * One will be used by the telemetry thread to process the completed
    //   measurements.
    // * One will be sitting in the channel, ready to be read in even if the
    //   telemetry thread is blocked while writing to the CSV file.
    for i in 0..=2 {
        let epoch = TelemetryEpoch::try_new()
            .with_context(|| format!("Failed to create telemetry epoch {i}"))?;
        sender
            .send(epoch)
            .with_context(|| format!("Failed to send telemetry epoch {i}"))?;
    }

    // Telemetry is written to a CSV file for later analysis.
    let csv_filename = format!("telemetry_{}.csv", stream_name);
    let mut csv = Csv::try_new(&csv_filename).with_context(|| {
        format!(
            "Failed to create CSV file for telemetry file '{}'",
            csv_filename
        )
    })?;

    thread::Builder::new()
        .name(format!("telemetry_{}", stream_name))
        .spawn(move || {
            let page_size =
                unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
            let mut last_allocated_bytes = 0;
            let mut last_allocation_count = 0;
            let mut last_freed = 0;

            while let Ok(mut epoch) = receiver.recv() {
                // If the `terminated` flag is set, we break out of the loop and
                // exit the telemetry thread gracefully.
                if epoch.terminated {
                    break;
                }

                // Now the inference thread is no longer updating the completed
                // epoch, we can save it.

                let current_allocated_bytes =
                    ALLOCATED_BYTES.load(Ordering::Relaxed);
                epoch.allocated_bytes = current_allocated_bytes
                    .saturating_sub(last_allocated_bytes);
                last_allocated_bytes = current_allocated_bytes;

                let current_allocation_count =
                    ALLOCATION_COUNT.load(Ordering::Relaxed);
                epoch.allocation_count = current_allocation_count
                    .saturating_sub(last_allocation_count);
                last_allocation_count = current_allocation_count;

                let current_freed_bytes = FREED_BYTES.load(Ordering::Relaxed);
                epoch.freed_bytes =
                    current_freed_bytes.saturating_sub(last_freed);
                last_freed = current_freed_bytes;

                let mut statm_file = std::fs::File::open("/proc/self/statm")
                    .expect("Failed to open /proc/self/statm");

                let mut sbuf = [0u8; 128];
                let n =
                    statm_file.read(&mut sbuf).expect("Failed to read statm");

                if n > 0 {
                    let mut start = 0;

                    // Ignore size.
                    while sbuf[start] != b' ' && start < n {
                        start += 1;
                    }
                    if start < n {
                        start += 1;
                    }

                    let mut end = start;
                    while sbuf[end] != b' ' && end < n {
                        end += 1;
                    }

                    let rss_str = std::str::from_utf8(&sbuf[start..end])
                        .expect("Invalid UTF-8 in statm");
                    let rss_pages: usize = rss_str.parse().expect(
                        "Failed to parse resident pages from /proc/self/statm",
                    );

                    epoch.rss_bytes = rss_pages * page_size;
                }

                epoch.fordblks_bytes =
                    unsafe { libc::mallinfo2().fordblks as usize };

                csv.write_record(&epoch)
                    .expect("Failed to write telemetry record to CSV");

                epoch.reset();
                sender
                    .send(epoch)
                    .expect("Failed to send clean telemetry epoch");
            }
        })
        .with_context(|| {
            format!("Failed to spawn telemetry thread for '{stream_name}'")
        })
}

File: code/pipeline-py/include/thread/telemetry.py
import csv
from dataclasses import dataclass
import resource
import sys
import threading
import time
from typing import Final

from hdrh.histogram import HdrHistogram

from include.os import Sender, Receiver, saturating_sub
import include.os.gc as gc


class TelemetryEpoch:
    """A telemetry epoch that contains six HdrHistogram instances for recording
    latency measurements in nanoseconds. Each histogram tracks latency for a
    specific stage of the inference pipeline, as well as the total latency."""

    UNBOUNDED_QUEUE_WAIT: int = 0
    """The index of the histogram for unbounded queue wait latency."""

    IDIOMATIC_QUEUE_WAIT: int = 1
    """The index of the histogram for idiomatic queue wait latency."""

    INFERENCE_EXEC: int = 2
    """The index of the histogram for inference latency."""

    MPSC_WAIT: int = 3
    """The index of the histogram for multi-producer single-consumer (MPSC)
    queue wait latency."""

    FUSION_EXEC: int = 4
    """The index of the histogram for fusion latency."""

    TOTAL_LATENCY: int = 5
    """The index of the histogram for total latency."""

    NUM_LATENCY_MEASURES: int = 6
    """The number of latency measures tracked in an epoch."""

    latency_nanos: list[HdrHistogram]
    """The histograms for recording latency measurements in nanoseconds for each
    stage of the inference pipeline, as well as the total latency."""

    lapped_frames: int = 0
    """The total number of frames lapped during the epoch."""

    dropped_frames: int = 0
    """The total number of frames dropped during the epoch."""

    gc_pause_ns: int = 0
    """The total time spent in garbage collection pauses during the epoch."""

    gc_blocks: int = 0
    """The total number of allocation blocks created during the epoch."""

    rss_bytes: int = 0
    """The total number of allocated RSS bytes during the epoch."""

    fan_pwm: int = 0
    """The current fan PWM value during the epoch."""

    terminated: bool = False
    """A flag indicating whether the telemetry thread should terminate."""

    def __init__(self):
        """Creates a new TelemetryEpoch with six HdrHistogram instances for
        recording latency measurements in nanoseconds. Each histogram is
        initialised to track values from 1 nanosecond to 60 seconds with 3
        significant figures of precision."""

        try:
            self.histograms = [
                HdrHistogram(1, 60_000_000_000, 3)
                for _ in range(self.NUM_LATENCY_MEASURES)
            ]
        except Exception as e:
            raise RuntimeError("Failed to create HdrHistogram for epoch") from e

        self.lapped_frames = 0
        self.dropped_frames = 0
        self.gc_pause_ns = 0
        self.gc_blocks = 0
        self.rss_bytes = 0

    def reset(self):
        """Resets the histograms and allocation statistics for the epoch."""

        for histogram in self.histograms:
            histogram.reset()

        self.lapped_frames = 0
        self.dropped_frames = 0
        self.gc_pause_ns = 0
        self.gc_blocks = 0
        self.rss_bytes = 0
        self.fan_pwm = 0


class TelemetryWriter:
    """Records latency measurements into an epoch buffer. Used by the inference
    thread to publish the results to the telemetry thread for processing, and to
    receive a fresh buffer for the next epoch."""

    SWAP_INTERVAL: int = 1_000_000_000
    """The interval at which the active epoch buffer is swapped with a fresh
    buffer received from the telemetry thread."""

    sender: Sender[TelemetryEpoch]
    """The sender channel for publishing the current epoch to the telemetry
    thread for processing."""

    receiver: Receiver[TelemetryEpoch]
    """The receiver channel for receiving a fresh epoch from the telemetry
    thread."""

    last_swap: int
    """The timestamp of the last buffer swap, used to determine when to next
    swap the buffers."""

    last_lapped_frames: int
    """The number of frames lapped since the last call to record, used to
    calculate the number of lapped frames for the current epoch."""

    last_dropped_frames: int
    """The number of frames dropped since the last call to record, used to
    calculate the number of dropped frames for the current epoch."""

    current_epoch: TelemetryEpoch
    """The currently active telemetry epoch for recording latency
    measurements."""

    is_terminated: bool
    """A flag indicating whether the telemetry thread has been signaled to
    terminate."""

    def __init__(
        self, sender: Sender[TelemetryEpoch], receiver: Receiver[TelemetryEpoch]
    ):
        """Constructs a new TelemetryWriter with the specified sender and
        receiver channels for double-buffered histogram processing.

        Args:
            sender: The sender channel for publishing the populated epoch to the
            telemetry thread for processing.
            receiver: The receiver channel for receiving a fresh epoch buffer
            from the telemetry thread."""

        self.sender = sender
        self.receiver = receiver
        self.last_swap = time.perf_counter_ns()
        self.last_lapped_frames = 0
        self.last_dropped_frames = 0
        self.current_epoch = receiver.receive()
        self.is_terminated = False

    def record(
        self, ts: list[int], lapped_frames: int, dropped_frames: int
    ) -> None:
        """Records latency measurements into the currently active telemetry
        epoch. If at least one second has elapsed since the last buffer swap,
        the active buffer is swapped with a fresh buffer received from the
        telemetry thread.

        Args:
            ts: An array of six timestamps in nanoseconds, representing the
            start and end times of each stage of the inference pipeline, as well
            as the total latency.
            lapped_frames: The total number of frames lapped.
            dropped_frames: The total number of frames dropped."""

        if self.is_terminated:
            return

        if time.perf_counter_ns() - self.last_swap >= self.SWAP_INTERVAL:
            self.swap_buffers()

        for i in range(TelemetryEpoch.NUM_LATENCY_MEASURES - 1):
            nanos = saturating_sub(ts[i + 1], ts[i])

            try:
                self.current_epoch.histograms[i].record_value(nanos)
            except Exception as e:
                raise RuntimeError(
                    f"Failed to record latency value {nanos} ns for measure {i}"
                ) from e

        total_nanos = saturating_sub(
            ts[TelemetryEpoch.TOTAL_LATENCY],
            ts[TelemetryEpoch.UNBOUNDED_QUEUE_WAIT],
        )

        try:
            self.current_epoch.histograms[
                TelemetryEpoch.TOTAL_LATENCY
            ].record_value(total_nanos)
        except Exception as e:
            raise RuntimeError(
                f"Failed to record total latency value {total_nanos} ns"
            ) from e

        newly_lapped = saturating_sub(lapped_frames, self.last_lapped_frames)
        self.current_epoch.lapped_frames += newly_lapped
        self.last_lapped_frames = lapped_frames

        newly_dropped = saturating_sub(dropped_frames, self.last_dropped_frames)
        self.current_epoch.dropped_frames += newly_dropped
        self.last_dropped_frames = dropped_frames

    def swap_buffers(self):
        """Swaps the active histogram buffer with the cleared buffer received
        from the background thread. The populated buffer is sent to the
        background thread for processing, and the timestamp of the last swap is
        updated."""

        """Try to receive the fresh epoch from the telemetry thread. If the
        channel is empty, we skip the swap and continue recording into the
        current epoch."""

        # Reset timer to guarantee one check per second.
        self.last_swap = time.perf_counter_ns()

        epoch = self.receiver.try_receive()

        if epoch != None:
            self.sender.send(self.current_epoch)
            self.current_epoch = epoch

    def terminate(self):
        """Signals the telemetry thread to terminate by setting the terminated
        flag in the current epoch and sending it to the telemetry thread. It
        needs to be guaranteed that the epoch is sent, regardless of whether the
        telemetry thread is blocked or not, to ensure that the telemetry thread
        can exit gracefully. Therefore swap_buffers() is not called here, and
        instead the termination epoch is sent directly."""

        if self.is_terminated:
            return

        self.is_terminated = True
        self.current_epoch.terminated = True
        self.sender.send(self.current_epoch)


class Csv:
    filename: str
    file: object
    writer: object

    def __init__(self, filename: str):
        self.filename = filename
        self.file = open(filename, 'w', newline='')
        self.writer = csv.writer(self.file)

        LABELS: Final[list[str]] = [
            'unbounded_wait',
            'idiomatic_wait',
            'inference_exec',
            'mpsc_wait',
            'fusion_exec',
            'total_latency',
        ]
        SUFFIXES: Final[list[str]] = ['p50', 'p99', 'p99_9', 'max']

        headings: list[str] = ['timestamp_ns']

        headings.extend(
            [f'{label}_{suffix}' for label in LABELS for suffix in SUFFIXES]
        )

        headings.extend(
            [
                'lapped_frames',
                'dropped_frames',
                'gc_pause_ns',
                'gc_blocks',
                'rss_bytes',
                'fan_pwm',
            ]
        )

        self.writer.writerow(headings)
        self.file.flush()

    def write_record(self, epoch: TelemetryEpoch, timestamp_ns: int):
        record: list[float] = [timestamp_ns]

        for i in range(TelemetryEpoch.NUM_LATENCY_MEASURES):
            histogram = epoch.histograms[i]
            record.append(histogram.get_value_at_percentile(50))
            record.append(histogram.get_value_at_percentile(99))
            record.append(histogram.get_value_at_percentile(99.9))
            record.append(histogram.get_max_value())

        record.extend(
            [
                epoch.lapped_frames,
                epoch.dropped_frames,
                epoch.gc_pause_ns,
                epoch.gc_blocks,
                epoch.rss_bytes,
                epoch.fan_pwm,
            ]
        )

        self.writer.writerow(record)
        self.file.flush()


def spawn_telemetry_thread(
    stream_name: str,
    sender: Sender[TelemetryEpoch],
    receiver: Receiver[TelemetryEpoch],
) -> threading.Thread:
    for i in range(3):
        epoch = TelemetryEpoch()
        sender.send(epoch)

    # Telemetry is written to a CSV file for later analysis.
    csv_filename = f"telemetry_{stream_name}.csv"

    try:
        csv = Csv(csv_filename)
    except Exception as e:
        raise RuntimeError(
            f"Failed to create CSV file for telemetry file '{csv_filename}'"
        ) from e

    def telemetry_thread():
        PAGE_SIZE: Final[int] = resource.getpagesize()

        last_gc_pause_ns = 0
        last_gc_blocks = 0

        while True:
            epoch = receiver.receive()
            timestamp_ns = time.time_ns()

            # If the `terminated` flag is set, we break out of the loop and exit
            # the telemetry thread gracefully.
            if epoch.terminated:
                break

            # Now the inference thread is no longer updating the completed
            # epoch, we can save it.

            current_gc_pause_ns = gc.pause_ns
            epoch.gc_pause_ns = saturating_sub(
                current_gc_pause_ns, last_gc_pause_ns
            )
            last_gc_pause_ns = current_gc_pause_ns

            current_gc_blocks = sys.getallocatedblocks()
            epoch.gc_blocks = saturating_sub(current_gc_blocks, last_gc_blocks)
            last_gc_blocks = current_gc_blocks

            try:
                with open('/proc/self/statm', 'r') as f:
                    rss_pages = int(f.read().split()[1])
                    epoch.rss_bytes = rss_pages * PAGE_SIZE
            except Exception:
                raise RuntimeError(
                    "Failed to read RSS bytes from /proc/self/statm"
                ) from e

            try:
                with open('/sys/class/hwmon/hwmon0/pwm1', 'r') as f:
                    epoch.fan_pwm = int(f.read().strip())
            except Exception:
                raise RuntimeError(
                    "Failed to read RSS bytes from /sys/class/hwmon/hwmon0/pwm1"
                ) from e

            try:
                csv.write_record(epoch, timestamp_ns)
            except Exception as e:
                raise RuntimeError(
                    f"Failed to write telemetry record to CSV file "
                    f"'{csv_filename}'"
                ) from e

            epoch.reset()

            try:
                sender.send(epoch)
            except Exception as e:
                raise RuntimeError(
                    "Failed to send clean telemetry epoch"
                ) from e

    try:
        thread = threading.Thread(
            target=telemetry_thread,
            name=f"telemetry_{stream_name}",
            daemon=True,
        )
        thread.start()
    except Exception as e:
        raise RuntimeError(
            f"Failed to spawn telemetry thread for '{stream_name}'"
        ) from e

    return thread
